libs/tokens/src/hooks/tokens/useSearchToken.ts

```diff
@@ -0,0 +1,154 @@
+import { useMemo } from 'react'
+import { useAtomValue } from 'jotai'
+import { activeTokensAtom, inactiveTokensAtom } from '../../state/tokens/allTokensAtom'
+import { useDebounce } from '@cowprotocol/common-hooks'
+import { useWeb3React } from '@web3-react/core'
+import { isAddress } from '@cowprotocol/common-utils'
+import ms from 'ms.macro'
+import { getTokenSearchFilter } from '../../utils/getTokenSearchFilter'
+import useSWR from 'swr'
+import { searchTokensInApi } from '../../services/searchTokensInApi'
+import { TokenWithLogo } from '@cowprotocol/common-const'
+import { environmentAtom } from '../../state/environmentAtom'
+import { parseTokensFromApi } from '../../utils/parseTokensFromApi'
+import { fetchTokenFromBlockchain } from '../../utils/fetchTokenFromBlockchain'
+import { tokenWithLogoFromToken } from '../../utils/tokenWithLogoFromToken'
+
+const IN_LISTS_DEBOUNCE_TIME = ms`100ms`
+const IN_EXTERNALS_DEBOUNCE_TIME = ms`1s`
+
+export type TokenSearchResponse = {
+  isLoading: boolean
+  blockchainResult: TokenWithLogo[] | null
+  externalApiResult: TokenWithLogo[] | null
+  activeListsResult: TokenWithLogo[] | null
+  inactiveListsResult: TokenWithLogo[] | null
+}
+
+type FromListsResult = {
+  tokensFromActiveLists: TokenWithLogo[]
+  tokensFromInactiveLists: TokenWithLogo[]
+}
+
+const emptyResponse: TokenSearchResponse = {
+  isLoading: false,
+  blockchainResult: null,
+  externalApiResult: null,
+  activeListsResult: null,
+  inactiveListsResult: null,
+}
+
+const emptyFromListsResult: FromListsResult = { tokensFromActiveLists: [], tokensFromInactiveLists: [] }
+
+export function useSearchToken(input: string | null): TokenSearchResponse {
+  const debouncedInputInList = useDebounce(input?.toLowerCase(), IN_LISTS_DEBOUNCE_TIME)
+  const debouncedInputInExternals = useDebounce(input?.toLowerCase(), IN_EXTERNALS_DEBOUNCE_TIME)
```

Comment:
Save `input?.toLowerCase()` to const to not do twice?

Reply:
Fixed: https://github.com/cowprotocol/cowswap/pull/3201/commits/c11ac54119cfe2701c61928feb8847e65a144c06